### ISSUE TITLE
RFE: Alternatives handled by RPM

### ISSUE THREAD

**Issue by Appadeia** [rpm-software-management/rpm#993]
Alternatives seem like a natural thing to be handled by RPM, but instead, it's handled by whatever a vendor cooks up for managing alternatives (which can get ugly specwise fast).

For packagers, alternatives managed by RPM could simply look like a `%alternative source dest` in the `%files` section of a package.

For how it could be implemented in rpm itself, files used as alternatives could be stored in the rpmdb, which rpm can hard/symlink from to the destination as the end user switches between them.

**Comment by Conan-Kudo** [comment]
I guess this would mean that it would be a transaction task to identify changes in alternatives and process them, similar to a transaction trigger?

**Comment by hellcp** [comment]
Generally, the amount of magic that has to be implemented in spec to have the alternatives makes me sad. As openSUSE distributions implements a lot of alternatives, in every aspect of the system, this makes a lot of specs just look awfully busy for no reason. This could also be beneficial from the scriplets standpoint.

This is typically what goes into adding alternatives:
```sh
Requires(post): update-alternatives
Requires(postun): update-alternatives

%install
mkdir -p %{buildroot}%{_sysconfdir}/alternatives
ln -s -f %{_sysconfdir}/alternatives/%{name} %{buildroot}%{link}

%post
update-alternatives --install %{link} %{name} %{path} %{priority}

%postun
if [ ! -f %{path} ] ; then
   update-alternatives --remove %{name} %{path}
fi

%files
%ghost %{_sysconfdir}/alternatives/%{name}
```

Which could be abbreviated with correct implementation into:
```sh
%files
%alternatives %{link} %{name} %{path} %{priority}
```

Which as it turns out, is way more readable ;)

**Comment by ffesti** [comment]
This is actually a nice use case to look how to make spec files more declarative and what road blocks are ahead. The main issue here is that there are all those moving pieces and rpm is not great in having them all in random order. Everything needs to go to the right place in the spec and so it is very difficult to expand this from a single place. Even more so if you what to have more than one %alternatives line.

So this starts with there only being one %files section per sub package and only one of each scriptlets and build scripts.
The %post and %postun scripts should actually done with file triggers but it's hard to pass data to them other than in the files themselves.

May be we need to address them one by one.

**Comment by hellcp** [comment]
Hm, a good question to ask maybe is if we are fine with how the alternatives work in the context of the system as is though. Maybe a more clever approach would be to manage alternatives entirely through rpm cli and rpmdb, without having to deal with the existing implementations, which obviously there are many of those, openSUSE distros and Fedora distros do not share the same implementation which implementing this via macro magic sounds like it would bring out the weirdest differences in both. It does sound like it would be easier to implement too, if the %alternative didn't have to have effect on what happens in %package, %install, %post and %postun, and only exist in %files, file triggers + rpm internals.

**Comment by pmatilai** [comment]
I agree, for rpm integration rpm should just take over the whole thing rather than try meddle with package scriptlets.

**Comment by pebolle** [comment]
@pmatilai: it's almost six years now, so I won't be surprised if you forgot most of this. Do you nevertheless still remember how you wanted to "take over the whole thing" here?

**Comment by pmatilai** [comment]
No, I don't.

Me agreeing with something isn't the same as wanting to do it, much less make it actually happen 😅

**Comment by pebolle** [comment]
(OK, that made me laugh.)

Perhaps the reason that nothing happened is because implementing this is, probably, a lot of work. And that's because alternatives is pretty complicated. (See for instance "auto" versus "manual".)

I guess I'll stick to making packages actually own the files they create in /etc/alternatives/ and /var/lib/alternatives/ to get a sense of the scope of the problem for the distro I happen to use. But, as I see no way to automate this, even that might be too much busywork.
